File: Python/textbook_track/chapter08/ba8c.py
# Implement the Lloyd Algorithm for k-Means Clustering
import functools as ft
import math
import sys


def convert_to_numbers(line, converter_func):
    return [converter_func(elem) for elem in line.split()]


def read_points(reader):
    return [convert_to_numbers(line, float) for line in reader]


def euclidean_distance(p, q):
    return math.sqrt(sum((p_i - q_i)**2 for p_i, q_i in zip(p, q)))


def select_closest_center(point, centers):
    calc_distance_from_point = ft.partial(euclidean_distance, q=point)
    distances = map(calc_distance_from_point, centers)
    min_ix, _ = min(enumerate(distances), key=lambda x: x[1])
    return min_ix


def assign_points_to_centers(points, centers):
    return [select_closest_center(point, centers) for point in points]


def recalculate_centers(points, center_assignment, k):
    centers = []
    for cluster_id in range(k):
        cluster_points = [point for point, ix in zip(points, center_assignment) if ix == cluster_id]
        cluster_size = len(cluster_points)
        center = [sum(coordinates) / cluster_size for coordinates in zip(*cluster_points)]
        centers.append(center)
    return centers


def are_same_centers(centers, updated_centers):
    return all(euclidean_distance(center, updated_center) <= 1e-12
               for center, updated_center in zip(centers, updated_centers))


def run_k_means_clustering(points, k):
    centers = points[:k]
    while True:
        center_assignment = assign_points_to_centers(points, centers)
        updated_centers = recalculate_centers(points, center_assignment, k)
        if are_same_centers(centers, updated_centers):
            return updated_centers

        centers = updated_centers


def main():
    reader = sys.stdin
    k, _ = convert_to_numbers(next(reader), int)
    points = read_points(reader)
    result = run_k_means_clustering(points, k)
    for point in result:
        print(' '.join(map(str, point)))


if __name__ == '__main__':
    main()

File: Scala/src/main/scala/TextbookTrack/Chapter08/BA8C.scala
package TextbookTrack.Chapter08

object BA8C {
  import scala.reflect.ClassTag
  import scala.annotation.tailrec

  type Point = Vector[Double]

  private def convertToNumber[T: ClassTag](line: String, converter: String => T): Vector[T] =
    line.split(" ").map(converter).toVector

  private def readParameters(line: String): (Int, Int) = convertToNumber(line, _.toInt) match {
    case Vector(k, m) => (k, m)
    case _ => throw new Exception("Unexpected number of parameters.")
  }

  private def euclideanDistance(p: Point, q: Point): Double =
    math.sqrt(p.lazyZip(q).foldLeft(0.0){ case (d, (pi, qi)) => d + math.pow(pi - qi, 2) })

  private def selectClosestCenter(point: Point, centers: List[Point]): Int =
    centers
      .iterator
      .map(euclideanDistance(point, _))
      .zipWithIndex
      .minBy{ case (distance, _) => distance }
      ._2

  private def assignPointsToCenters(points: List[Point], centers: List[Point]): List[Int] =
    points.map(selectClosestCenter(_, centers))

  private def recalculateCenters(points: List[Point], centerAssignment: List[Int], k: Int): List[Point] =
    (0 until k).map{
      clusterId =>
        val clusterPoints: List[Point] =
          points
            .lazyZip(centerAssignment)
            .withFilter{ case (_, ix) => ix == clusterId }
            .map{ case (point, _) => point }
            .toList
        val clusterSize: Int = clusterPoints.length
        clusterPoints.transpose.map(_.sum / clusterSize).toVector
    }.toList

  private def areSameCenters(centers: List[Point], updatedCenters: List[Point]): Boolean =
    centers.lazyZip(updatedCenters).forall{euclideanDistance(_, _) <= 1e-12 }

  def runKMeansClustering(points: List[Point], k: Int): List[Point] = {
    @tailrec
    def loop(centers: List[Point]): List[Point] = {
      val centerAssignment: List[Int] = assignPointsToCenters(points, centers)
      val updatedCenters: List[Point] = recalculateCenters(points, centerAssignment, k)
      if (areSameCenters(centers, updatedCenters)) updatedCenters
      else loop(updatedCenters)
    }

    loop(points.take(k))
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val (k, _): (Int, Int) = readParameters(reader.next())
    val points: List[Point] = reader.map(convertToNumber(_, _.toDouble)).toList
    val result: List[Point] = runKMeansClustering(points, k)
    result.foreach(point => println(point.mkString(" ")))
  }
}
